src/icp.1
.\"                                      -*- nroff -*-
.\" icp.1 - Manual page for icp.
.\"
.\" Copyright (C) 2001, 2002, 2004, 2005 Oskar Liljeblad
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU Library General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH ICP "1" "August 12, 2005" "icp (renameutils)"
.SH NAME
icp \- Copy a file by editing the destination name using GNU readline
.SH SYNOPSIS
.B icp
.RI [ OPTION ] " FILE"...
.SH DESCRIPTION
This manual page document describes the \fBicp\fP command.

\fBicp\fP is a program to interactively copy a single file.
It does this by allowing the file name to be edited inline
with GNU readline. This is very similar to using \fBcp\fP(1) and
editing the filename on the shell command-line, with one
exception - the filename does not have to be typed twice.

The \fBicp\fP program normally executes \fBcp\fP(1) to do the actual renaming.
This can however be changed with the \-\-command option.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').

All options except those listed below are passed to \fBcp\fP.
\fBcp\fP.
.TP
\fB\-\-command\fR=\fIFILE\fR
Specify command to run instead of the default `cp'.
You do not need to specify the directory name of FILE if FILE
is found in the current path (as set by the $PATH environment
variable).

It is assumed that the command specified accepts the same set
of long and short
options that require an argument as cp. If not, you should
specify those options with \-\-arg\-options (see below).
\-\-arg-options is not needed as long as you specify option and
option value in a single word, like `--suffix=bak' instead of
`--suffix bak' when passing extra options to cp via icp.

.TP
\fB\-\-arg\-options\fR=\fIOPTION\fR[,\fIOPTION\fR...]
Specify what options for cp (or the command specified with
\-\-command) require an argument. Options may be short
(e.g. \-S) as well as long (e.g. \-\-suffix).
The default list is `S,V,no-preserve,sparse,suffix,version-control'
(correct for cp of GNU Coreutils 5.2.1).

This list is necessary because icp needs to know what arguments
specified on the command line are files to copy, or option
values (following an option).
.TP
\fB\-\-help\fR
Show summary of options.
.TP
\fB\-\-version\fR
Output version information and exit.
.SH EXAMPLES
Using icp as a Bash alias for cp:
.br
	\fBalias cp=icp\fR
.SH REPORTING BUGS
Report bugs to <\fIoskar@osk.mine.nu\fP>.
.SH SEE ALSO
\fBcp\fP(1)
.SH AUTHOR
The author of \fBrenameutils\fP and this manual page is Oskar Liljeblad <\fIoskar@osk.mine.nu\fP>.
.SH COPYRIGHT
Copyright \(co 2001, 2002, 2004, 2005 Oskar Liljeblad

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


NEWS
2005-08-23: renameutils 0.8.1 released.
  Now with complete French translation.

2005-08-23: renameutils 0.8.0 released.
  Removed debugging statements rendering imv and icp unusable (oops!)
  (reported by Francois Marier).
  Use mv(1) to perform the actual rename. Fixes Debian bug 316113
  (http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=316113). (Reported by
  Daniel Hulme through Francois Marier.)
  Improved some messages and error messages.
  Change order of inclusion of readline headers, fixing complation problem
  on some distributions (patch from Brian Hall).
  Added --command and --arg-options option to icmd (icp and imv).

2005-08-16: web site updated.
  Add some missing entries to this history.

2005-08-16: renameutils 0.7.0 released
  Added Swedish translation.
  Added French translation (from Francois Marier).
  The executables in the RPM packages now links with ncurses to prevent
  problems on Red Hat and Fedora systems.
  Handle POSIXLY_CORRECT in imv (fixes rename bug) (from Brian Hall).
  Change imv to pass command line through to mv if at least two file names
  are specified (or --target-directory option for mv) (suggested by
  Francois Marier).
  imv now checks if the source file exists prior to asking for destination
  name (suggested by Francois Marier).
  imv now detects and handles options requiring arguments for mv.
  imv now trims the edited file name. This means that the destination file
  name cannot have leading or trailing white space.
  Merge code base for imv and icp into icmd.
  Got rid of year ranges as recommended by the FSF.
  Improve the manual pages a little.
  Add Gnulib modules quote, dirname. Remove stat as it no longer exists in
  Gnulib.
  Fixed some typos in messages of the program.

2005-06-28: web site updated.
  Update description in README and web page.

2005-06-28: renameutils 0.6.0 released.
  Added new programs, icp and qcp (suggested by Mario Vanoni).
  Quote file names and strings before printing them (like GNU coreutils).
  Make use of more gnulib modules (quotearg, yesno, version-etc).
  Fixed a serious memory management-related bug in imv causing renaming to
  fail for no reason.
  Fixed a potential segmentation fault when printing a particular error
  message.
  Added copyright and license header to source files and manual pages.
  Manual page updates:
    Added manual page for deurlname.
    Corrected grammar.
    Removed section in imv about placing it in the Midnight Commander menu.
  Gettext is now external (no intl directory).
  Use Gnulib properly (modules.gnulib and autogen.sh removed).
  Added debian package.
  Added MANIFEST.sources file.
  Fix potential bug in cross-reference solution code.
  Added documentation for qcmd --help from Stephen Kolupaev, thanks.

2004-09-17: renameutils 0.5.1 released.
  Clarified README about requirements.
  Include Makefile.am.head in lib and m4 directories in distribution
  tarball.
  Minor clean ups to the autogen.sh, Makefile.am and configure.ac files.
  Now compiles with --disable-nls as well:
    Don't include (<gettext.h is enough).
    Include <locale.h> where setlocale is used.
  Check that locale was set correctly.

2004-08-26: renameutils 0.5.0 released.
  Added deurlname - a program to remove URL-escaped character sequences
  from filenames (such as %20 for space).
  Internationalized renameutils - added support for gettext.
  Removed debian support since renameutils has been in the official Debian
  archives for quite some time.
  Removed show-control-chars variable and option (automaticly assumed since
  we are using gettext).
  Fixed a bug in dual-column editformat that would cause a segfault when
  renaming files with trailing space in their names (reported by Chris
  Jantzen).
  Fixed a bug in qmv that made it impossible to rename files larger than
  2GB (reported by Chris Gentle).
  Corrected makefiles to allow for creation of staticly linked executables.
  renameutils now requires a C89-compatible compiler and C library.
  Modules from gnulib are used to improve compatibility.

2002-08-07: renameutils 0.4.0 released.
  Enable autowidth for dual-column by default, for real this time.
  Fixed "Illegal instruction" crash with '-fdc -oseparate'.
  Added more checks for headers and functions to the configure script.

2002-01-15: renameutils 0.3.0 released.
  Added new option `autowidth' to the dual-column edit format, which is
  enabled by default.
  Handle case when the edited file can't be read correctly (for whatever
  reason).
  Fixed bug in imv that made it impossible to overwrite files.
  Now compiles with more error checking (compile with -Wall).
  Fixed some compilation errors.
  Added files for making an RPM package of renameutils.

2001-10-25: renameutils 0.2.0 released.
  Added manual pages for qmv and imv.
  Added files for making a renameutils Debian package.
  The author has changed email-address to oskar@osk.mine.nu.

2001-08-14: renameutils 0.1.2 released.
  Made it possible to compile with readline 4.1 as well as 4.2.
  Added note in README about requirements.

2001-08-13: renameutils 0.1.1 released.
  Fixed a subtle bug regarding order of #includes that would cause build to
  fail in some occasions.

2001-08-13: renameutils 0.1.0 released.
  First public release.


src/icp.c
/* icp.c - Interactive cp(1).
 *
 * Copyright (C) 2001-2005 Oskar Liljeblad
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#if HAVE_CONFIG_H
#include <config.h>
#endif
#if HAVE_UNISTD_H
#include <unistd.h> 	    	    /* POSIX */
#endif
#include <stdlib.h> 	    	    /* C89 */
#include <signal.h> 	    	    /* C89 */
#include <stdio.h>  	    	    /* C89 */
#include <errno.h>  	    	    /* C89 */
#include <locale.h> 	    	    /* C89 */
#include <stdbool.h>	    	    /* Gnulib (POSIX) */
#include <progname.h>	    	    /* Gnulib */
#include <quotearg.h>	    	    /* Gnulib */
#include <yesno.h>  	    	    /* Gnulib */
#include <version-etc.h>    	    /* Gnulib */
#include <gettext.h>	    	    /* Gnulib (gettext) */
#define _(String) gettext(String)
#define N_(String) (String)
#include <readline/readline.h>	    /* Readline */
#include <readline/history.h>	    /* Readline */
#include "xalloc.h"
#include "common/error.h"
#include "common/io-utils.h"
#include "common/string-utils.h"
#include "common/common.h"

#define PROGRAM "icp"

static char *sourcefile = NULL;
static char *first_text = NULL;
const char version_etc_copyright[] = N_("Copyright (C) 2001-2004 Oskar Liljeblad");

static int
insert_first_text(void)
{
    return rl_insert_text(first_text);
}

static void
int_signal_handler(int signal)
{
    puts("");
    warn(_("no changes made"));
    exit(0);
}

static void
display_imv_help(void)
{
    printf(_("Usage: %s [OPTION]... FILE\n\
Copy a file by editing the destination name using GNU readline.\n\
All options except the options listed below are passed to cp(1).\n\
\n\
      --help                 display this help and exit\n\
      --version              output version information and exit\n\
\n\
Report bugs to <%s>.\n"), program_name, PACKAGE_BUGREPORT);
}

int
main(int argc, char **argv)
{
    struct sigaction action;
    char *newname;
    bool gotend = false;
    int c;
    int len;

    set_program_name(argv[0]);

    if (setlocale (LC_ALL, "") == NULL)
    	die(_("invalid locale"));
    if (bindtextdomain (PACKAGE, LOCALEDIR) == NULL)
    	die_errno(NULL);
    if (textdomain (PACKAGE) == NULL)
    	die_errno(NULL);

    if (argc == 2 && strcmp(argv[1], "--help") == 0) {
	display_imv_help();
	exit(0);
    }
    if (argc == 2 && strcmp(argv[1], "--version") == 0) {
	/*version_etc_copyright = N_("Copyright (C) 2001-2004 Oskar Liljeblad");*/
	version_etc(stdout, PROGRAM, PACKAGE, VERSION, "Oskar Liljeblad", NULL);
	exit(0);
    }

    memset(&action, 0, sizeof(sigaction));
    action.sa_handler = int_signal_handler;
    action.sa_flags = SA_RESTART;
    if (sigaction(SIGINT, &action, NULL) < 0)
	die_errno(NULL);
    action.sa_handler = SIG_IGN;
    if (sigaction(SIGQUIT, &action, NULL) < 0)
	die_errno(NULL);

    for (c = 1; c < argc; c++) {
	if (!gotend && strcmp(argv[c], "--") == 0) {
	    gotend = true;
	} else if (gotend || argv[c][0] != '-') {
	    if (sourcefile != NULL)
		die(_("too many arguments"));
	    sourcefile = argv[c];
	}
    }

    if (sourcefile == NULL)
	die(_("missing file argument"));

    len = strlen(sourcefile);
    if (sourcefile[len-1] == '/')
	sourcefile[len-1] = '\0';

    first_text = xstrdup(sourcefile);
    do {
	add_history(sourcefile);
	rl_startup_hook = insert_first_text;
	newname = readline("> ");

	if (newname != NULL
		&& strcmp(newname, "") != 0
		&& strcmp(newname, sourcefile) != 0) {
	    char **args;

	    if (file_exists(newname)) {
		printf(_("%s: overwrite `%s'? "), program_name, quotearg(newname));
		fflush(stdout);
		if (!yesno()) {
		    free(first_text);
		    first_text = newname;
		    continue;
		}
	    }

	    args = xmalloc(sizeof(char *) * (argc + (gotend ? 2 : 3)));
	    memcpy(args, argv, argc * sizeof(char *));

	    if (gotend) {
		args[argc] = newname;
		args[argc+1] = NULL;
	    } else {
		args[argc] = "--";
		args[argc+1] = newname;
		args[argc+2] = NULL;
	    }

	    args[0] = "cp";
	    if (execvp(args[0], args) < 0)
	    	die_errno(_("cannot execute %s"), args[0]);
	    die(_("cannot execute %s"), args[0]);
	}
    } while (false);

    if (newname == NULL)
	puts("");

    warn(_("no changes made"));
    exit(0);
}
